File: ios/Classes/InappPurchasePlugin.swift
import Flutter
import UIKit
import StoreKit
import Foundation
import Combine

public class InappPurchasePlugin: NSObject, FlutterPlugin {
    
    private static let store = Store()
    
    private static let availableSubscriptionStreamHandler = SubscriptionsStreamHandler()
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let methodChannel = FlutterMethodChannel(
            name: "com.lanars.inapp_purchase/methods",
            binaryMessenger: registrar.messenger()
        )
        let subscriptionsChannel = FlutterEventChannel(
            name: "com.lanars.inapp_purchase/subscriptions",
            binaryMessenger: registrar.messenger()
        )
        let purchasedSubscriptionsChannel = FlutterEventChannel(
            name: "com.lanars.inapp_purchase/purchased_subs",
            binaryMessenger: registrar.messenger()
        )
        let instance = InappPurchasePlugin()
        registrar.addMethodCallDelegate(instance, channel: methodChannel)
        subscriptionsChannel.setStreamHandler(self.availableSubscriptionStreamHandler)
        purchasedSubscriptionsChannel.setStreamHandler(PurchasedSubscriptionStreamHandler())
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let method = Method(rawValue: call.method) else {
            result(FlutterMethodNotImplemented)
            return
        }
        
        switch method {
        case .refreshProducts:
            onRefreshProducts(call, result)
        case .buy:
            onBuyProduct(call, result)
        case .getPlatformVersion:
            result("iOS " + UIDevice.current.systemVersion)
        }
    }
    
    private func onRefreshProducts(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        print("Refresh products")
        Task {
            await InappPurchasePlugin.store.requestProducts()
        }
        result(nil)
    }
    
    private func onBuyProduct(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        if let args = BuyArguments(call) {
            print("Buy product \(args.productId)")
            Task {
                do {
                    let product = InappPurchasePlugin.store.subscriptions.first { product in
                        product.id == args.productId
                    }
                    if let product = product {
                        Task {
                            try await InappPurchasePlugin.store.purchase(product)
                        }
                    }
                    result(nil)
                } catch {
                    result(nil)
                }
            }
        }
        result(nil)
    }
    
    class SubscriptionsStreamHandler : NSObject, FlutterStreamHandler {
        private var eventSink: FlutterEventSink?
        
        private var cancellabe: Cancellable?
        
        func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
            self.eventSink = events
            startListeningPurchases()
            return nil
        }
        
        func onCancel(withArguments arguments: Any?) -> FlutterError? {
            self.eventSink = nil
            cancellabe?.cancel()
            cancellabe = nil
            return nil
        }
        
        func sendEvent(_ event: Any) {
            eventSink?(event)
        }
        
        private func startListeningPurchases() {
            cancellabe = store.$subscriptions.sink { products in
                let jsonProducts = products.map { $0.toJsonString() }.filter { $0 != nil }
                self.sendEvent(jsonProducts)
            }
        }
    }
    
    class PurchasedSubscriptionStreamHandler : NSObject, FlutterStreamHandler {
        private var eventSink: FlutterEventSink?
        
        private var cancellable: Cancellable?
        
        func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
            self.eventSink = events
            startListeningPurchases()
            return nil
        }
        
        func onCancel(withArguments arguments: Any?) -> FlutterError? {
            self.eventSink = nil
            cancellable?.cancel()
            cancellable = nil
            return nil
        }
        
        func sendEvent(_ event: Any) {
            eventSink?(event)
        }
        
        private func startListeningPurchases() {
            cancellable = store.$purchasedSubscriptions.sink { products in
                let jsonProducts = products.map { $0.toJsonString() }.filter { $0 != nil }
                self.sendEvent(jsonProducts)
            }
        }
    }
}

enum Method: String {
    case refreshProducts = "refreshProducts()"
    case buy = "buy(Product)"
    case getPlatformVersion = "getPlatformVersion" // TODO: remove
}

struct BuyArguments {
    let productId: String
    
    init?(_ call: FlutterMethodCall) {
        if let args = call.arguments as? [String: Any],
           let productId = args["productId"] as? String {
            self.productId = productId
        } else {
            return nil
        }
    }
}

extension Product: Encodable {
    enum CodingKeys: CodingKey {
        case id, title, description, price, displayPrice
    }
    
    public func encode(to encoder: Encoder) throws {
        var container: KeyedEncodingContainer<Product.CodingKeys> = encoder.container(keyedBy: CodingKeys.self)
        
        try container.encode(id, forKey: .id)
        try container.encode(displayName, forKey: .title)
        try container.encode(description, forKey: .description)
        try container.encode(price, forKey: .price)
        try container.encode(displayPrice, forKey: .displayPrice)
    }
}

extension Product {
    func toJsonString() -> String? {
        if let data = try? JSONEncoder().encode(self) {
            return String(data: data, encoding: .utf8)
        }
        return nil
    }
}

File: android/src/main/kotlin/com/lanars/inapp_purchase/InappPurchasePlugin.kt
package com.lanars.inapp_purchase

import android.app.Activity
import android.util.Log
import com.android.billingclient.api.BillingClient
import com.android.billingclient.api.ProductDetails
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch
import org.json.JSONObject
import java.lang.ref.WeakReference
import kotlin.coroutines.resume
import kotlin.coroutines.suspendCoroutine

private const val TAG = "InAppPurchasePlugin"

/** InAppPurchasePlugin */
class InAppPurchasePlugin : FlutterPlugin, MethodCallHandler, ActivityAware {

    private lateinit var billingManager: BillingManager2
    private lateinit var methodChannel: MethodChannel
    private lateinit var subscriptionsChannel: EventChannel
    private lateinit var purchasedSubscriptionsChannel: EventChannel

    private val coroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.Main)

    private var weakActivity: WeakReference<Activity> = WeakReference(null)

    private val purchaseVerifier = PurchaseVerifier { purchaseToken, productId ->
        return@PurchaseVerifier suspendCoroutine { continuation ->
            methodChannel.invokeMethod(
                "verifyPurchase",
                mapOf(
                    "purchaseToken" to purchaseToken,
                    "productId" to productId
                ),
                object : MethodChannel.Result {
                    override fun success(result: Any?) {
                        continuation.resume(result as? Boolean ?: false)
                    }

                    override fun error(
                        errorCode: String,
                        errorMessage: String?,
                        errorDetails: Any?
                    ) {
                        continuation.resume(false)
                    }

                    override fun notImplemented() {
                        continuation.resume(false)
                    }
                }
            )
        }
    }

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        billingManager = BillingManager2(
            flutterPluginBinding.applicationContext,
            coroutineScope,
            purchaseVerifier
        )
        methodChannel = MethodChannel(flutterPluginBinding.binaryMessenger, METHOD_CHANNEL_NAME)
        methodChannel.setMethodCallHandler(this)
        subscriptionsChannel = EventChannel(
            flutterPluginBinding.binaryMessenger,
            SUBSCRIPTIONS_CHANNEL_NAME
        )
        subscriptionsChannel.setStreamHandler(SubscriptionStreamHandler())
        purchasedSubscriptionsChannel = EventChannel(
            flutterPluginBinding.binaryMessenger,
            PURCHASED_SUBSCRIPTIONS_CHANNEL_NAME
        )
        purchasedSubscriptionsChannel.setStreamHandler(PurchasedSubscriptionStreamHandler())
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        val method = Method.fromMethodName(call.method) ?: run {
            result.notImplemented()
            return
        }

        when (method) {
            Method.REFRESH_PRODUCTS -> {
                Log.d(TAG, "onMethodCall: ${method.methodName}")

                val identifiers = call.argument<List<String>>("identifiers") ?: run {
                    result.error("identifiers is null", null, null)
                    return
                }
                coroutineScope.launch(Dispatchers.IO) {
                    try {
                        billingManager.refreshProducts(identifiers)
                        result.success(null)
                    } catch (e: Throwable) {
                        result.error(e.localizedMessage.orEmpty(), null, null)
                    }
                }
            }

            Method.BUY_PRODUCT -> {
                Log.d(TAG, "onMethodCall: ${method.methodName}")

                val activity = weakActivity.get() ?: run {
                    result.error("activity is null", null, null)
                    return
                }
                val productId = call.argument<String>("productId") ?: run {
                    result.error("productId is null", null, null)
                    return
                }
                coroutineScope.launch {
                    val product = billingManager.products.first().firstOrNull {
                        it.productId == productId
                    } ?: run {
                        result.error("product not found", null, null)
                        return@launch
                    }
                    billingManager.launchPurchase(
                        activity,
                        product,
                        product.subscriptionOfferDetails!!.first().offerToken
                    )
                    result.success(null)
                }
            }

            Method.RESTORE_PURCHASES -> {
                // TODO: implement
                Log.d(TAG, "onMethodCall: ${method.methodName}")
                result.notImplemented()
            }

            Method.GET_PLATFORM_VERSION -> {
                result.success("Android ${android.os.Build.VERSION.RELEASE}")
            }
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        coroutineScope.cancel()
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        this.weakActivity = WeakReference(binding.activity)
    }

    override fun onDetachedFromActivityForConfigChanges() {
        this.weakActivity.clear()
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        this.weakActivity = WeakReference(binding.activity)
    }

    override fun onDetachedFromActivity() {
        this.weakActivity.clear()
    }

    companion object {
        const val METHOD_CHANNEL_NAME = "com.lanars.inapp_purchase/methods"
        const val SUBSCRIPTIONS_CHANNEL_NAME = "com.lanars.inapp_purchase/subscriptions"
        const val PURCHASED_SUBSCRIPTIONS_CHANNEL_NAME = "com.lanars.inapp_purchase/purchased_subs"
    }

    enum class Method(val methodName: String) {
        REFRESH_PRODUCTS("requestProducts([String])"),
        BUY_PRODUCT("buy(Product)"),
        RESTORE_PURCHASES("restorePurchases()"),
        GET_PLATFORM_VERSION("getPlatformVersion"); // TODO: remove

        companion object {
            fun fromMethodName(methodName: String): Method? {
                return values().firstOrNull { it.methodName == methodName }
            }
        }
    }

    inner class SubscriptionStreamHandler : EventChannel.StreamHandler {
        private var eventSink: EventChannel.EventSink? = null
        private var listenJob: Job? = null

        override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
            this.eventSink = events
            listenJob = coroutineScope.launch {
                billingManager.subscriptions.collect { subscriptions ->
                    sendEvent(subscriptions.map { it.toJsonString() })
                }
            }
        }

        override fun onCancel(arguments: Any?) {
            this.eventSink = null
            listenJob?.cancel()
        }

        private fun sendEvent(event: Any) {
            eventSink?.success(event)
        }
    }

    inner class PurchasedSubscriptionStreamHandler : EventChannel.StreamHandler {
        private var eventSink: EventChannel.EventSink? = null
        private var listenJob: Job? = null

        override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
            this.eventSink = events
            listenJob = coroutineScope.launch {
                billingManager.purchasedProducts.collect { products ->
                    sendEvent(products.map { it.toJsonString() })
                }
            }
        }

        override fun onCancel(arguments: Any?) {
            this.eventSink = null
            listenJob?.cancel()
        }

        private fun sendEvent(event: Any) {
            eventSink?.success(event)
        }
    }
}

fun ProductDetails.toJsonString(): String {
    val json = JSONObject().apply {
        put("id", productId)
        put("title", title)
        put("description", description)
        put(
            "price",
            subscriptionOfferDetails!!.first().pricingPhases.pricingPhaseList.first().priceAmountMicros / 1000000.0
        )
        put(
            "displayPrice",
            subscriptionOfferDetails!!.first().pricingPhases.pricingPhaseList.first().formattedPrice
        )
        put(
            "type",
            when (productType) {
                BillingClient.ProductType.SUBS -> "autoRenewable"
                BillingClient.ProductType.INAPP -> "consumable" // TODO: handle non-consumable
                else -> ""
            }
        )
    }
    return json.toString()
}
